### PR TITLE
upstream: add degraded hosts and update LB to prefer healthy hosts over degraded

### PR DESCRIPTION
**Pull request by snowp**
This implements the load balancing part of supporting degraded hosts.
    Works by maintaining separate priority loads for degraded hosts, and
    selecting degraded hosts whenever we spill over from the healthy hosts.

Locality weights are similarly also maintained separately for degraded
    hosts. This ensures that we scale locality weights by the size of hosts
    under consideration when routing.

*Risk Level*: High, significant refactors to core load balancing code
*Testing*: n/a for now
*Docs Changes*: n/a for now
*Release Notes*: n/a for now
[Optional Fixes #Issue]
[Optional *Deprecated*:]

Part of #5063

### CODE REVIEW

File: include/envoy/upstream/upstream.h [hunk 1]
```diff
@@ -109,6 +109,11 @@ class Host : virtual public HostDescription {
    */
   virtual bool healthy() const PURE;
 
+  /**
+   * @return whether in aggregate a host is degraded and should be deprioritized during routing.
+   */
+  virtual bool degraded() const PURE;
```
Comment (rshriram) [inline include/envoy/upstream/upstream.h]: how is this conceptually different from outlier detection where a host fails repeatedly (we have success threshold/failure threshold, and other signals).
If the criteria for a degraded host is same as those gleaned from other means (failing health checks or from outlier detection), then wont it be easier to provide an API-ish way to indicate that priority routing should takeinto account failing hosts as well (albeit at lower priority)

Comment (rshriram) [reply include/envoy/upstream/upstream.h]: looking at the PR, I guess thats what you are doing? the PR description seemed to hint at an API level change (which I mistook for data plane api changes

Comment (snowp) [reply include/envoy/upstream/upstream.h]: Yeah, maybe I should improve the title. 

The main bit in this PR is really the LB changes that makes it prioritize healthy hosts before degraded hosts. A lot of the changes end up being related to extracting the list of degraded hosts and passing those around due to the APIs changed are used all over the place.

**Comment by snowp** [comment]
@dio Mind giving this a look? Previous offer of splitting this up still stands

**Comment by dio** [comment]
@snowp let me take a look. Meanwhile, let's merge master and fix the conflicting files 😄

File: source/common/upstream/thread_aware_lb_impl.cc [hunk 2]
```diff
@@ -57,7 +60,13 @@ ThreadAwareLoadBalancerBase::LoadBalancerImpl::chooseHost(LoadBalancerContext* c
   }
   const uint64_t h = hash ? hash.value() : random_.random();
 
-  const uint32_t priority = LoadBalancerBase::choosePriority(h, *per_priority_load_);
+  // TODO(snowp): Since we don't keep track of whether we selected from a dregraded priority
```
Comment (dio) [inline source/common/upstream/thread_aware_lb_impl.cc]: nit: s/dregraded/degraded

File: source/common/upstream/upstream_impl.h [hunk 3]
```diff
@@ -194,6 +194,7 @@ class HostImpl : public HostDescriptionImpl,
     outlier_detector_ = std::move(outlier_detector);
   }
   bool healthy() const override { return !health_flags_; }
+  bool degraded() const override { return false; }
```
Comment (dio) [inline source/common/upstream/upstream_impl.h]: Sorry, so we also override this somewhere else?

**Comment by dio** [comment]
@snowp let's close this in favor of: #5202?

**Comment by snowp** [comment]
Sure

**Pull request closed without merging**